// functions/cheque_ft/addEdit.js
//
//
// by : mostafa

var convert_date = require('date');
var mongoose = require('mongoose');
var db = require('db');

module.exports = function(data, cb) {

  cb = cb || function() {};

  if (!data) {
    cb({
      status: 400,
      message: "low args"
    });
  } else {
    data = String.remove_empty_data(data);

    validate(data, function(status) {
      switch (status.valid) {
        case true:
          data = status.data;
          if (data) {
            //add
            // console.open(data);
            data.type = "cheque_ft";
            data.items = [];
            data.items.push({
              type: "cheque"
            });
            data.items[0].serial = data.serial
            var _date = data.date.split("/");
            convert_date.to_miladi({
              year: _date[0],
              month: _date[1],
              day: _date[2]
            }, function(miladi) {
              data.items[0].date = miladi
            })
            data.items[0].account_number = data.account_number
            data.items[0].branch_id = data.branch_id
            data.items[0].price = data.price
            if (data.comment) {
              data.items[0].comment = data.comment
            };
            db.operations.findOne({
              $and: [{
                "items.serial": data.serial
              }, {
                type: "cheque_ft"
              }]
            }, {
              _id: true
            }).lean().exec(function(err, cheque_exists) {
              if (err) {
                cb({
                  status: 500,
                  error_obj: err
                });
              } else if (cheque_exists) {
                cb(null, {
                  exists: true
                });
              } else {
                add(data, function(err, status) {
                  cb(err, status);
                });
              }
            });
          }
          break;
        case false:
          cb(null, {
            add: false,
            not_valid: true
          })
          break;
      }
    });

  }
};


function add(data, cb) {
  cb = cb || function() {};
  //add_cheque_ft
  new db.operations(data).save(function(err) {
    if (err) {
      cb({
        status: 400,
        error_obj: err
      }, null);
    } else {
      cb(null, {
        add: true
      });
    }
  })
  //
};

function validate(data, cb) {

  cb = cb || function() {};


  if (data &&
    data.person_id &&
    data.serial &&
    data.date &&
    data.account_number &&
    data.branch_id &&
    data.price
  ) {
    data.serial = String.NumtoEn(data.serial);
    data.account_number = String.editPersianCh(data.account_number);
    data.price = Number.NumtoEn(data.price);
    if (data.comment) {
      data.comment = String.editPersianCh(data.comment);
    };
    //
    cb({
      data: data,
      valid: true
    });
  } else {
    cb({
      valid: false
    });
  }
};
