// functions/bankaccount/remaining.js
//
//
// by : mostafa
"use strict";

const mongoose = require('mongoose');
const db = require('db');
// const encrypt = require('encrypt');
// const date_to_persian = require('functions/date_to_persian');

module.exports = function(data, cb) {

  cb = cb || function() {};
  if (!data) {
    cb({
      status: 400,
      message: "low args"
    });
  } else {
    //
    // console.open(data);
    var match = {
      $match: {
        "items.bankacc_id": mongoose.Types.ObjectId(data)
      }
    };
    var unwind = {
      $unwind: '$items'
    };
    var match2 = {
      $match: {
        $or: [
          {
            $and: [{
              $or: [{
                type: "cheque_ft"
              }, {
                "items.type": "cheque"
              }]
            },{
              "items.bankacc_id": mongoose.Types.ObjectId(data)
            },{
              "items.status": "ph"
            }]
          },{
            $and: [{
              $and: [{
                type: {
                  $ne: "cheque_ft"
                }
              }, {
                "items.type": {
                  $ne: "cheque"
                }
              }]            },{
              "items.bankacc_id": mongoose.Types.ObjectId(data)
            }]
          }
        ]
      }
    }
    var group = {
      $group: {
        _id: '$items.bankacc_id',
        sum: {
          $sum: '$items.price'

        }
      }
    };
    var project = {
      $project: {
        _id: false,
        bankacc_id: '$_id.bankacc_id',
        // sum: '$sum'
        remaining: '$sum'
      }
    };

    var agg = [];

    agg.push(match);
    agg.push(unwind);
    agg.push(match2);
    agg.push(group);
    agg.push(project);

    db.operations.aggregate(
      agg
    ).exec(function(err, result) {
      if (err) console.log(err);
      if (result) {
        return cb(null, result[0]);
        // db.persons.populate(result, {
        //   path: 'person_id',
        //   select: 'first_name last_name id credit_limit'
        // }, function(err, info) {
        //   if (err) throw err;
        //   if (info) {
        //     // console.open(info);
        //     return cb(null, info[0])
        //     // res.json(info);
        //   }
        // });
      }
    });

  }

}
