// functions/by/addEdit.js
//
//
var mongoose = require('mongoose');
var db = require('db');


module.exports = function(by, cb) {

  cb = cb || function() {};

  if (!by && by.title && by.title != null) {
    cb({
      status: 400,
      message: "low args"
    });
  } else {
    by.title = String.editPersianCh(by.title);
    //
    if (by && by.person_id && !by._id) {
      //add
      by = String.remove_empty_data(by);

      db.bys.findOne({
        title: by.title,
        person_id: by.person_id
      }, {
        _id: true
      }).lean().exec(function(err, by_exists) {
        if (err) {
          cb({
            status: 500,
            error_obj: err
          });
        } else if (by_exists) {
          cb(null, {
            exists: true
          });
        } else {
          add(by, function(err, status) {
            cb(err, status);
          });
        }
      });
    } else if (by && by.person_id && by._id) {
      //edit
      // edit(by, function(err, status) {
      //   cb(err, status);
      // });

    }

  }
}


function add(by, cb) {
  cb = cb || function() {};

  var new_id = mongoose.Types.ObjectId();
  // console.log("_id => " + id);
  by._id = new_id;
  //add_by
  new db.bys(by).save(function(err) {
    // error with db
    if (err) {
      cb({
        status: 400,
        error_obj: err
      }, null);
    } else {
      // save
      cb(null, {
        add: true,
        by_id : new_id
      });
    }
  });
}

function edit(by, cb) {

  cb = cb || function() {};

  db.bys.update({
    _id: by._id
  }, {
    $set: by
  }, function(err, res) {
    if (err) {
      cb(null, {
        error: true,
        type: 'edit'
      })
    } else if (res) {
      cb(null, {
        edit: true
      });
    }
  });
  //end function
}
