// db/schemas/persons.js
//
//
//
"use strict";

const mongoose = require('mongoose');
const validate = require('mongoose-validator');

const Schema = mongoose.Schema;
const ObjectId = Schema.Types.ObjectId;

const nameValidator = [
  validate({
    validator: 'isLength',
    arguments: [3, 20],
    message: 'Name should be between {ARGS[0]} and {ARGS[1]} characters'
  })
];

const persons_schema = Schema({
  //show on tbl : id, type, group_name, name,  mandeh, mahiat
  id: Number,
  type: {
    type: String,
    enum: ['person', 'company', 'shareholder']
  },
  first_name: {
    type: String,
    trim: true
      // ,
      // required : true,
      // validate : nameValidator
  },
  last_name: {
    type: String,
    trim: true
  },
  gender: Boolean,
  group_id: {
    type: ObjectId,
    ref: 'g_persons'
  },
  credit_limit: Number,
  address: String,
  comment: String,
  details: [{
    type: {
      type: String,
      enum: ["mobile", "work", "home", "address", "bankacc", "creditcart", "sheba", "other"]
    },
    number: String,
    comment: String
  }],
  creator_id: {
    type: ObjectId,
    ref: 'users'
  },
  disable: Boolean
});


module.exports = mongoose.model('persons', persons_schema);
