// db/schemas/users.js
//
//
//

var mongoose = require('mongoose');
var Schema = mongoose.Schema;
var ObjectId = Schema.Types.ObjectId;

// these schema is for users, every one want's to use this app is a user
// and there is not diffrece between students and employees ,everyone!
// tasks are deffrence of any job.
var users_schema = Schema({
    //
    first_name: {
       type: String,
       trim: true
    },
    last_name: {
       type: String,
       trim: true
    },
   // male or female -> true or not exists
   gender: Boolean,
   //null
   mobile: {
      type: String,
      trim: true,
   },
   // username
   email: {
      type: String,
      lowercase: true,
      trim: true
   },
   // encypted password
   password: String,

   //_id of g_users
   group_id: [{
      type: ObjectId,
      ref: 'g_users'
   }],

   // comment about this users, every thing needed..
   comments: String,

   // _id of the user add this user
   creator_id: {
      type: ObjectId,
      ref: 'users'
   },

   // if true , user is disable
   _disable: Boolean

});


module.exports = mongoose.model('users', users_schema);
